common/archiver/gcloud/historyArchiver.go

```diff
@@ -0,0 +1,402 @@
+// The URI identifies the resource from which history should be accessed and it is up to the implementor to interpret this URI.
+// This method should thrift errors - see filestore as an example.
+func (h *historyArchiver) Get(ctx context.Context, URI archiver.URI, request *archiver.GetHistoryRequest) (*archiver.GetHistoryResponse, error) {
+
+	err := h.validateURI(ctx, URI)
+	if err != nil {
+		return nil, &shared.BadRequestError{Message: archiver.ErrInvalidURI.Error()}
+	}
+
+	if err := archiver.ValidateGetRequest(request); err != nil {
+		return nil, &shared.BadRequestError{Message: archiver.ErrInvalidGetHistoryRequest.Error()}
+	}
+
+	var token *getHistoryToken
+	if request.NextPageToken != nil {
+		token, err = deserializeGetHistoryToken(request.NextPageToken)
+		if err != nil {
+			return nil, &shared.BadRequestError{Message: archiver.ErrNextPageTokenCorrupted.Error()}
+		}
+	} else if request.CloseFailoverVersion != nil {
+		_, historyhighestPart, historyCurrentPart, err := h.getHighestVersion(ctx, URI, request)
+		if err != nil {
+			return nil, &shared.InternalServiceError{Message: err.Error()}
+		}
+		token = &getHistoryToken{
+			CloseFailoverVersion: *request.CloseFailoverVersion,
+			HighestPart:          *historyhighestPart,
+			CurrentPart:          *historyCurrentPart,
+			NextBatchIdx:         0,
+		}
+	} else {
+		highestVersion, historyhighestPart, historyCurrentPart, err := h.getHighestVersion(ctx, URI, request)
+		if err != nil {
+			return nil, &shared.InternalServiceError{Message: err.Error()}
+		}
+		token = &getHistoryToken{
+			CloseFailoverVersion: *highestVersion,
+			HighestPart:          *historyhighestPart,
+			CurrentPart:          *historyCurrentPart,
+			NextBatchIdx:         0,
+		}
+	}
+
+	response := &archiver.GetHistoryResponse{}
+	numOfEvents := 0
+	numOfBatches := 0
+
+	for partNum := token.CurrentPart; partNum <= token.HighestPart; partNum++ {
+
+		historyBatches := []*shared.History{}
+		filename := constructHistoryFilenameMultipart(request.DomainID, request.WorkflowID, request.RunID, token.CloseFailoverVersion, partNum)
+		encodedHistoryBatches, err := h.gcloudStorage.Get(ctx, URI, filename)
+
+		if err != nil {
+			return nil, &shared.InternalServiceError{Message: err.Error()}
+		}
+
+		if encodedHistoryBatches == nil {
+			return nil, &shared.InternalServiceError{Message: "Fail retrieving history file: " + URI.String() + "/" + filename}
+		}
+
+		batches, err := decodeHistoryBatches(encodedHistoryBatches)
+		if err != nil {
+			return nil, &shared.InternalServiceError{Message: err.Error()}
+		}
+		historyBatches = append(historyBatches, batches...)
+		historyBatches = historyBatches[token.NextBatchIdx:]
+
+		for _, batch := range historyBatches {
+			response.HistoryBatches = append(response.HistoryBatches, batch)
+			numOfBatches++
+			numOfEvents += len(batch.Events)
+
+		}
+
+		if numOfEvents >= request.PageSize {
```

Comment:
I think this check should be done inside the for loop above.

Reply:
done!